proposals/0292-package-registry-service.md

```diff
@@ -796,7 +791,8 @@ and attempt to reuse those credentials to impersonate the user.
 ```
 
 This kind of attack can be mitigated on an individual basis
-by adding `.swiftpm/config` to a project's `.gitignore` file.
+by adding `.swiftpm/config` to a project's `.gitignore` file
+or storing credentials in a `.netrc` file.
```

Comment:
I believe we prefer to actively discourage embedding credentials in the configuration and even validate the URLs to not having them.

cc @lukasa

Reply:
As [discussed in the 2nd review thread](https://forums.swift.org/t/se-0292-2nd-review-package-registry-service/46917/35), we won't be able to remove support for hardcoded credentials until `.netrc` files are supported on all platforms.

Reply:
as mentioned in https://forums.swift.org/t/se-0292-2nd-review-package-registry-service/46917/39, I think in the proposal should reflect the ideal security design and we deal with the implementation aspect in the implementation phase - especially in this case where the implementation gap is small.

Reply:
also cc @mmarston

Reply:
I agree — I just see the question of how credentials are stored to be an implementation detail for this proposal. 

We currently support hardcoded URL credentials on all platforms and `.netrc` on newer versions of macOS, and this proposal doesn't change anything about that. The proposal correctly calls this out as a security consideration, but it doesn't require an immediate solution.

I would push back on the idea that the implementation gap is small, though. Beyond adding support for `.netrc` on Linux (and Windows), I think there are a few different ways that you could approach the tooling as far as migrating existing use of hardcoded credentials, warning against them, or automatically converting hardcoded credentials and managing `.netrc` files. I think it'd be useful to consider the entire scope of the problem independently as a separate feature.

Reply:
I think it is critical from a security perspective to avoid giving developers even the slightest temptation to place credentials in their repository. While I agree that it is possible for developers to mitigate this threat, the widespread and consistent pattern of credential disclosure via git repositories is something that I think we need to be minimising.

In this specific instance, there are no existing hardcoded credentials because the registries feature has not yet shipped. This means we don’t yet have to solve the question of migrating them. There are _other_ places in SwiftPM where credentials may be hardcoded and migration does represent an issue there, but it’s separate to this proposal.

Reply:
> While I agree that it is possible for developers to mitigate this threat, the widespread and consistent pattern of credential disclosure via git repositories is something that I think we need to be minimising.

We can take immediate steps to mitigate this for all projects in the future by adding `.swiftpm/config` to the default `.gitignore` file in  new projects. I just opened a PR for this: https://github.com/apple/swift-package-manager/pull/3468

I don't believe there's a strong consensus among the security community that having token credentials in the userinfo component of a URL is an inherently bad thing. `user:password`? Yes, that's a bad idea. But scoped access credentials continue to be used by GitHub and many other service providers (to wit, a [Google Trends comparison of ".env" to ".netrc"](https://trends.google.com/trends/explore?date=all&geo=US&q=%22.env%22,%22.netrc%22)). As a developer, I often prefer URL credentials to external configuration, and I think it'd be a mistake to remove this feature outright.

Or at the very least, I don't think this proposal needs to state a strong preference for one method of authentication over another. For the purposes of resolving external package dependencies through a registry service, this is an implementation detail — changing that particular behavior doesn't substantively affect the rest of the proposal.

Reply:
> We can take immediate steps to mitigate this for all projects in the future by adding .swiftpm/config to the default .gitignore file in new projects. I just opened a PR for this: apple/swift-package-manager#3468

This is a good change to make, thanks!

> I don't believe there's a strong consensus among the security community that having token credentials in the userinfo component of a URL is an inherently bad thing. We can all agree that exposing your username and password in plaintext is a bad idea.

Having a username/password in the userinfo portion of a URL is not inherently a bad idea either. What is important is that password-equivalent credentials should be kept out of repositories. It is important that users are not encouraged to place password equivalent credentials into files that they may check in to source control. There very much is consensus in the security community that that's a bad idea.

I don't believe I've proposed anywhere removing URL credentials. What I have proposed, consistently, is that we should not accept URLs with userinfo components in any file that you might be checking into source control (i.e. within your repository).

Reply:
> Or at the very least, I don't think this proposal needs to state a strong preference for one method of authentication over another. For the purposes of resolving external package dependencies through a registry service, this is an implementation detail — changing that particular behavior doesn't substantively affect the rest of the proposal.

I think this is also entirely reasonable: if this pitch simply wants to remain silent on whether you can add userinfo credentials that seems totally fair.

Reply:
> What is important is that password-equivalent credentials should be kept out of repositories.

I also think that it is important that if URLs are used in credentials then the URL should be sanitized before it is included in any log/diagnostic output.

> We can take immediate steps to mitigate this for all projects in the future by adding .swiftpm/config to the default .gitignore file in new projects.

It is worth noting that I see scenarios where developers would want some project-level `.swiftpm/config` settings to be committed. In particular, when using a private registry in a team setting I can see value in committing the `registries` settings (assuming credentials are managed separately). With npm, we have users that configure their project-level `.npmrc` file with a private CodeArtifact registry and they commit this to source control. That way when another developer on the team runs `git clone` they'll automatically be set up to use the correct registry for that project.

(I've previously pointed this out in [this comment](https://github.com/mattt/swift-evolution/commit/07497a799f80ecb80bea163b84e6f31cc331bc91#commitcomment-48329368))

> I think this is also entirely reasonable: if this pitch simply wants to remain silent on whether you can add userinfo credentials that seems totally fair.

I'm OK if  `SE-0292` doesn't pitch an improvement on credential management, but I'd like to see it addressed in a future Swift evolution proposal. The part of `SE-0292` that would be affected is the new `swift package-registry set` command. I would envision additional arguments to provide credentials.

Reply:
> I also think that it is important that if URLs are used in credentials then the URL should be sanitized before it is included in any log/diagnostic output.

I agree. This is discussed in the [Security](https://github.com/apple/swift-evolution/blob/ac89166961deb77d125c451a9db9f7ace9b7d10b/proposals/0292-package-registry-service.md#information-disclosure) section.

> It is worth noting that I see scenarios where developers would want some project-level .swiftpm/config settings to be committed.

As I mentioned in https://github.com/apple/swift-package-manager/pull/3468#issuecomment-832205120, ignoring `.swiftpm/config` by default doesn't preclude it from being added to version control.